utils/logalloc.cc

```diff
@@ -1910,18 +1910,27 @@ void tracker::impl::full_compaction() {
 }
 
 static void reclaim_from_evictable(region::impl& r, size_t target_mem_in_use, is_preemptible preempt) {
-    while (true) {
-        auto deficit = shard_segment_pool.total_memory_in_use() - target_mem_in_use;
-        auto occupancy = r.occupancy();
-        auto used = occupancy.used_space();
-        if (used == 0) {
-            break;
+    llogger.debug("reclaim_from_evictable: total_memory_in_use={} target={}", shard_segment_pool.total_memory_in_use(), target_mem_in_use);
+
+    // Before attempting segment compaction, try to evict at least deficit and one segment more so that
+    // for workloads in which eviction order matches allocation order we will reclaim full segments
+    // without needing to perform expensive compaction.
+    auto deficit = shard_segment_pool.total_memory_in_use() - target_mem_in_use;
+    auto used = r.occupancy().used_space();
+    auto used_target = used - std::min(used, deficit + segment::size);
+
+    while (shard_segment_pool.total_memory_in_use() > target_mem_in_use) {
+        if (preempt && need_preempt()) {
+            llogger.debug("reclaim_from_evictable preempted");
+            return;
+        }
```

Comment:
We typically put preemption checks at the end, to guarantee forward progress. Otherwise a very slow (or contended) cpu can lead to this function always being preempted, without doing any work.

Reply:
It doesn't matter in this case, since even if we don't check for preemption here, we will check for it after evicting a single row.
Progress here is guaranteed by the fact that if background reclaim works too slowly, then eventually reclaim will be called without preemption, and do the work.

Reply:
Hm. That at least deserves a comment, because it looks like a bug.

Reply:
Fixed. I have both moved the check further down and added a comment.